src/L1VotePool.sol

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IGovernor} from "openzeppelin/governance/Governor.sol";
+import {IWormhole} from "wormhole/interfaces/IWormhole.sol";
+
+import {WormholeReceiver} from "src/WormholeReceiver.sol";
+
+contract L1VotePool is WormholeReceiver {
+  /// @notice The address of the L1 Governor contract.
+  IGovernor public governor;
+
+  /// @dev Thrown when proposal does not exist.
+  error MissingProposal();
+
+  /// @dev Thrown when a proposal vote is invalid.
+  error InvalidProposalVote();
+
+  /// @dev Contains the distribution of a proposal vote.
+  struct ProposalVote {
+    uint128 inFavor;
+    uint128 against;
+    uint128 abstain;
+  }
+
+  /// @notice A mapping of proposal id to the proposal vote distribution.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @param _core The address of the Wormhole Core contract.
+  /// @param _governor The address of the L1 Governor contract.
+  constructor(address _core, address _governor) WormholeReceiver(_core) {
+    governor = IGovernor(_governor);
+  }
+
+  /// @notice Receives a message from L2 and saves the proposal vote distribution.
+  /// @param encodedMsg The encoded message Wormhole VAA from the L2.
+  function receiveEncodedMsg(bytes memory encodedMsg) public override {
+    (IWormhole.VM memory vm,,) = _validateMessage(encodedMsg);
+
+    (uint256 proposalId, uint128 against, uint128 inFavor, uint128 abstain) =
+      abi.decode(vm.payload, (uint256, uint128, uint128, uint128));
+
+    ProposalVote memory existingProposalVote = proposalVotes[proposalId];
+    if (
+      existingProposalVote.against <= against || existingProposalVote.inFavor <= inFavor
+        || existingProposalVote.abstain <= abstain
+    ) revert InvalidProposalVote();
+
+    // Save proposal vote
+    proposalVotes[proposalId] = ProposalVote(inFavor, against, abstain);
+
+    _castVote(
+      proposalId,
+      ProposalVote(
+        inFavor - existingProposalVote.inFavor,
+        against - existingProposalVote.against,
+        abstain - existingProposalVote.abstain
+      )
+    );
+  }
+
+  /// @notice Casts vote to the L1 Governor.
+  /// @param proposalId The id of the proposal being cast.
+  function _castVote(uint256 proposalId, ProposalVote memory vote) internal {
+    if ((vote.against + vote.inFavor + vote.abstain) <= 0) revert MissingProposal();
```

Comment:
Wait, why is the error message for this condition `MissingProposal`? In fact, why are we checking this condition at all?

Reply:
Hm yes, this was leftover from the `castVote` refactor, removing

Reply:
https://github.com/ScopeLift/l2-flexible-voting/pull/9/commits/15bd0e5f6d18523e967ee6f93499a706679e9a5b